luarules/gadgets/unit_xp_range_bonus.lua

```diff
@@ -14,20 +14,26 @@ if not gadgetHandler:IsSyncedCode() then
     return false
 end
 
-
 local SetUnitWeaponState = Spring.SetUnitWeaponState
 local GetUnitExperience = Spring.GetUnitExperience
 local SetUnitMaxRange = Spring.SetUnitMaxRange
 local GetUnitDefID = Spring.GetUnitDefID
 
 local updateList = {}
+local XPDefs = {}
+
+for unitDefID, unitDef in pairs(UnitDefs) do
+	if unitDef.customParams ~= nil then
+		if unitDef.customParams.rangexpscale ~= nil then
+			XPDefs[unitDefID] = unitDef
```

Comment:
cache just the thing you need, the unitDef is a large proxy:
```suggestion
			XPDefs[unitDefID] = unitDef.customParams.rangexpscale
```
(and use accordingly later)

Reply:
If `rangexpscale` is the only thing you use from the unitDef yes.

Reply:
I need the ID as I need unit def, have implemented above feedback from badosu

Reply:
Looking at the whole file, it seems to me you don't need the unitDef. The only thing you use is the `rangexpscale`. So you could save a bit more perf by just caching it instead.

Reply:
Ah, nevermind, you use the weapon, I found another issue there but it's outside the scope of this issue.

You can cache a pair of `{ unitDef.customParams.rangexpscale, WeaponDefs[unitDef.weapons[1].weaponDef].range }` instead. Then access it as 

```lua
local rangeXPScale, originalRange =  XPDefs[unitDefID]
```